### DISCUSSION TITLE
Symfony 6.4 WebTestCase emty body

### DISCUSSION THREAD

**Discussion by D3xime** [api-platform/core#6521] [Q&A] [answered]
Hello everyone, 

i am having some problems with phpUnit WebTestCase, i want to test a route created on my own with POST method, but when i try to send a body i got an error that saying an empty body is sent which is false because i set a body in my test. I have added the link to the stackoverflow questions i have posted. I really need help, i can't get over that and i tried to ask ClaudeAi ChatGPT but none have the solutions for my problems. if you are more details answer me i will be happy to respond to your question. 

https://stackoverflow.com/questions/78871456/symfony-6-4-webtestcase-emty-body

In hope you can help me, thanks in advance.

**Comment by jamesisaac** [comment]
Hi D3xime,

I'm not sure where you're getting that signature from of calling `$client->request` with 6 arguments.  You can see the function signature of `HttpClientInterface:;request` in Symfony 6.4 here: https://github.com/symfony/symfony/blob/7.2/src/Symfony/Contracts/HttpClient/HttpClientInterface.php#L85 , and it only has 3 arguments.

As per the docs: https://symfony.com/doc/6.4/http_client.html#uploading-data , it also has built-in JSON support, so try like this:

```php
    $client->request(
        'POST',
        '/api/zone/1',
        [
            'headers' => ['Accept' => 'application/json'],
            'json' => $parameters,
        ]
    );
```

I also notice you wrote the header as `'CONTENT_TYPE' => 'application/json',` .  The header name should be with a hyphen, not an underscore: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Type

**Reply by D3xime** [reply]
Thanks for the reply, but like i explained, i use this `client->request` in a Test file, i got it by simply extending WebTestCase and doing `$client = static::createClient();`. It's not a part of HttpClientInterface, and I don't know if i can use this class in test. 
The code provided below works well in symfony 5.4 because i adapted from an old project.


here all my class to have the full context : 
```php
class ZoneControllerTest extends WebTestCase
{
    public function testChangeStateOnZoneWithInvalidData(): void
    {
        $client = static::createClient();

        $parameters = []; // Corps de la requête JSON vide

        $client->request(
            'POST',
            '/api/zone/1',
            [],  // Pas de paramètres GET
            [],  // Pas de fichiers
            [
                'CONTENT_TYPE' => 'application/json',
                'ACCEPT' => 'application/json',
                'CONTENT_LENGTH' => strlen('{"light":"On"}')
            ],
            '{"light":"On"}'  // Corps de la requête JSON
        );
        dump($client->getRequest()->getContent());
        $response = $client->getResponse();

        // Afficher la réponse pour le débogage
        // dump($response->getContent());   s

        // Vérifiez si la réponse est un JSON valide
        $this->assertJson($response->getContent(), 'Response is not valid JSON');

        // Vérifiez le code de statut
        $this->assertEquals(400, $response->getStatusCode(), 'Expected status code 400');
    }
}
```

**Reply by jamesisaac** [reply, marked as answer]
Oh I see.  Seeing as you're using APIP, it's probably worth extending ApiTestCase rather than WebTestCase? https://github.com/api-platform/core/blob/main/src/Symfony/Bundle/Test/ApiTestCase.php#L37 that would use HttpClientInterface, and have other benefits as described in the APIP docs: https://api-platform.com/docs/core/testing/

I'm not particularly familiar with Symfony's BrowserKit, but if you want to keep using it instead, a couple of things I notice:
1. It has a `jsonRequest` method that would simplify things: https://symfony.com/doc/current/components/browser_kit.html#making-requests
2. Seeing as Content-Type is an http header, looks like you'd need to prefix it with HTTP_: https://github.com/symfony/symfony/blob/7.1/src/Symfony/Component/BrowserKit/AbstractBrowser.php#L322

**Reply by D3xime** [reply]
I didn't know about ApiTestCase, i will use it instead of the WebTestCase for testing my Api. 
I just find out, i am not using a route created by Api platform. Sorry for inconvenience.
